File: likelee-ui/src/utils/pricingDefaults.ts
export const MIN_BASE_MONTHLY_CENTS = 15000;
export const DEFAULT_PRICING_GRACE_SECONDS = 60;

export function parseIsoToMs(value?: string | null): number | null {
  if (!value) {
    return null;
  }
  const hasTimezone = /[zZ]|[+-]\d{2}:\d{2}$/.test(value);
  const normalized = hasTimezone ? value : `${value}Z`;
  const parsed = Date.parse(normalized);
  return Number.isNaN(parsed) ? null : parsed;
}

export function isDefaultPricing(profile: {
  base_monthly_price_cents?: number | null;
  created_at?: string | null;
  pricing_updated_at?: string | null;
}): boolean {
  const monthly = profile.base_monthly_price_cents ?? null;

  const matchesMin = monthly === MIN_BASE_MONTHLY_CENTS;
  if (!matchesMin) {
    return false;
  }

  const createdAt = parseIsoToMs(profile.created_at ?? null);
  const pricingUpdatedAt = parseIsoToMs(profile.pricing_updated_at ?? null);
  if (!pricingUpdatedAt || !createdAt) {
    return true;
  }

  return (pricingUpdatedAt - createdAt) / 1000 <= DEFAULT_PRICING_GRACE_SECONDS;
}

export function shouldDefaultVisibilityOn(profile: {
  public_profile_visible?: boolean | null;
  visibility?: string | null;
  base_monthly_price_cents?: number | null;
  created_at?: string | null;
  pricing_updated_at?: string | null;
}): boolean {
  const publicVisible = profile.public_profile_visible;
  const visibility = (profile.visibility ?? "").trim().toLowerCase();

  if (publicVisible !== false) {
    return false;
  }
  if (!(visibility === "" || visibility === "private")) {
    return false;
  }
  return isDefaultPricing(profile);
}

File: likelee-server/src/pricing_defaults.rs
const MIN_BASE_MONTHLY_CENTS: i64 = 15_000;
const DEFAULT_PRICING_GRACE_SECONDS: i64 = 60;

fn parse_rfc3339(value: Option<&str>) -> Option<chrono::DateTime<chrono::FixedOffset>> {
    value.and_then(|v| chrono::DateTime::parse_from_rfc3339(v).ok())
}

pub fn is_default_pricing(row: &serde_json::Value) -> bool {
    let monthly = row.get("base_monthly_price_cents").and_then(|v| v.as_i64());
    let matches_min = monthly == Some(MIN_BASE_MONTHLY_CENTS);
    if !matches_min {
        return false;
    }

    let created_at = parse_rfc3339(row.get("created_at").and_then(|v| v.as_str()));
    let pricing_updated_at = parse_rfc3339(row.get("pricing_updated_at").and_then(|v| v.as_str()));
    if pricing_updated_at.is_none() {
        return true;
    }
    match (created_at, pricing_updated_at) {
        (Some(created), Some(pricing)) => {
            (pricing - created).num_seconds() <= DEFAULT_PRICING_GRACE_SECONDS
        }
        _ => true,
    }
}

pub fn should_default_visibility_on(row: &serde_json::Value) -> bool {
    let public_visible = row.get("public_profile_visible").and_then(|v| v.as_bool());
    let visibility = row
        .get("visibility")
        .and_then(|v| v.as_str())
        .unwrap_or("")
        .trim()
        .to_lowercase();

    if public_visible != Some(false) {
        return false;
    }
    if !(visibility.is_empty() || visibility == "private") {
        return false;
    }
    if !is_default_pricing(row) {
        return false;
    }

    let created_at = parse_rfc3339(row.get("created_at").and_then(|v| v.as_str()));
    let updated_at = parse_rfc3339(row.get("updated_at").and_then(|v| v.as_str()));

    // Privacy-safe: if timestamps are missing or unparseable, default to false (keep private)
    match (created_at, updated_at) {
        (Some(created), Some(updated)) => updated == created,
        _ => false, // Conservative default: don't enable visibility when timestamps are missing
    }
}
